### ISSUE TITLE
[Discussion] The state of optimus-manager

### ISSUE THREAD

**Issue by Askannz** [Askannz/optimus-manager#543]
Hi all,

This is @Askannz, creator and current maintainer of optimus-manager. I just wanted to update users about the state of this project.

## TLDR

* I lack the expertise to keep developing optimus-manager, and that's why it hasn't been actively developed for a while
* My test hardware is pretty old at this point and doesn't reflect modern Optimus laptops
* I'm looking to hand over the project to a new maintainer, if anyone is interested
* In the meantime I'll keep the project working to the extent that I can

## The long version

First off, apologies for the lack of updates for the past…err…couple years? I wish I could blame it on the COVID mayhem but the truth is that I don’t feel qualified to work on this project anymore.

### Dealing with bug reports

I originally started optimus-manager because I had an Optimus laptop I wanted to run Archlinux on, and realized Archlinux didn’t have as good a solution as Ubuntu for handling that kind of dual-GPU setup. So I copied Ubuntu’s solution, made a script for myself and published it to Github thinking that maybe others would find it useful. I never imagined this project would become so popular, and I just can’t believe it’s gotten over 2k stars on Github. I’m proud and thankful for it, but frankly also a bit overwhelmed.

Optimus-manager has never been anything more than a bunch of existing scripts cobbled together into a nicer package, and to this day the way it does GPU switching is still pretty hacky (dynamically re-generating Xorg configs, hooking into display managers, calling undocumented ACPI functions…). I think this process works well enough for most users, but it has a lot of moving parts, many of which are still black boxes that I don’t understand very well. My professional skills are mostly in MLops with a sprinkle of data science, so I can only take educated guesses as to what’s going on within the guts of X11, display managers, or the Nvidia driver.

This makes dealing with Github issues pretty frustrating. When people report that optimus-manager doesn’t work on their particular configuration, there is simply not that much I can do. It doesn’t help that problems can have similar symptoms (black screens, freezing, or incorrect GPU switching) but completely different causes, requiring a lot of triage and case-by-case handling. That’s also compounded by quirks of individual brand/models of laptops, by people running optimus-manager on distribution that I don’t use and can’t test (Manjaro), by the capriciousness of the display stack on Linux, or by the fact that it’s difficult to get logs out of a user when the bug they’re trying to report takes down their entire desktop.

Several times I tried to take a day to go through the reports one by one, comb through logs, identify the offending part of the stack, and at least try to offer solutions but it was so absurdly time-consuming I stopped doing it.

## Keeping up with the hardware

The good news is that the official support for Optimus on Linux seems to have gotten a little better since I started this project, with recent laptops supporting some kind of power management without having to unload the entire driver. But that brings me to my next point, which is that I don’t have the hardware required to keep up with those advancements.

When I started developing optimus-manager I had 2 Optimus laptops to test it on, each from a different brand and with slightly different hardware configurations. That was convenient because I could try new versions of optimus-manager on both, with different Arch-based distributions, and be reasonably confident it was going to work on other laptops.
Back to the present day, one of those laptops is long gone and the other one isn’t doing much better (rather literally falling apart). Furthermore, it only has a Pascal-generation GPU (GTX1060) which is almost 7 years old and doesn’t support the new power management methods added by Nvidia, so I have to rely on users to test the support for those. I don’t even use this laptop daily anymore, having upgraded to a proper desktop station now that I’m done with the student life and don’t have to move around so much.

## What now

Because I only use my Optimus laptop very rarely now, optimus-manager is in the unfortunate situation where its creator (me) doesn’t even dogfood it. At this point I’m worried that some day, some update somewhere in the stack is going to break things and I won’t know about it or be able to fix it. Or Nvidia will update their power management again in a way that requires optimus-manager to adapt.

There’s also the question of Wayland support, which people have been asking for for some time. At this point X11 is clearly on the way out, but Wayland is such a different beast I wouldn’t even know where to begin. I suppose I could copy Ubuntu’s implementation once more, but it would likely need a significant rewrite and again I would only be able to test it on my aging hardware.

So if I can’t actively develop optimus-manager, where does that leave the project?

To be honest, I don’t have a good idea of how many people are using optimus-manager today. Newer Optimus laptops have the alternative option of leaving the Nvidia driver running and letting it handle the power management (the equivalent of “hybrid” mode in optimus-manager), which makes optimus-manager redundant in that case. But the Github issues are still active and the AUR package is still reasonably popular so I assume this project is still useful for some folks. That’s why I don’t want to just let it die.

So for now I’ll keep monitoring issues, updating the AUR packages (including the gdm-prime rebase) and if there are low-risk PRs submitted to Github I’ll try to review them. That’s not much of a change over the last few years, but that means you shouldn’t expect large updates or new features for the near future.

For the longer term (assuming optimus-manager is still relevant by then), the right thing to do would be to hand this project to someone who has the hardware and the skills necessary to pick it up (and to be honest I should have done this a long time ago rather than letting Github issues pile up).

If you feel up to the task, please get in touch (either here or at my email which you can find in the PKGBUILD). We can discuss how to do the handover: I can add you as a collaborator to this repository, or archive it and add a redirect to your fork.

**Comment by ugoenyioha** [comment]
thank you for all the hard work you did on this project. despite the note, I want to be the first to acknowledge how many persons you helped with this project and the impact it had helping us get our laptops working on Linux. thanks again.

**Comment by mike-lloyd03** [comment]
Thank you for your hard work in providing Optimus-Manager to the community. This tool made Nvidia GPUs and good battery life at the same time possible. I am mostly on Wayland now but I need X11 for some apps still and I depend on Optimus-Manager in those situations. 

Whichever way you go, best of luck to you and thank you again.

**Comment by LukeFernandes** [comment]
Thanks very much for taking the time to write this update, and for all your work on this. It's really appreciated and as @shashanknimje says, without it we wouldn't be able to use Arch/its variants as our daily drivers. I'm not sure there's an alternative yet (bumblebee doesn't really work and hasn't for a long time), especially for those of us on Pascal or older. So I think it's important this project doesn't die. Occasional maintenance from you is very appreciated. Of course it would be great if someone could jump in to maintain it :). I don't feel knowledgeable enough to do that.

**Comment by VorpalBlade** [comment]
Thank you for this existing at all.

>  But the Github issues are still active and the AUR package is still reasonably popular so I assume this project is still useful for some folks. That’s why I don’t want to just let it die.

Yeah, my old Thinkpad T480 with a MX150 still works perfectly fine and I use it daily. I don't plan to upgrade until it falls apart or is unusably slow, which will likely be at least 5 more years.  I don't do anything 3D heavy on it, in fact I keep it in intel-only mode 99% of the time (which is thankfully possible since on this laptop the external display outputs are connected directly to the intel GPU). Still, optimus-manager is there the few times I need something more. Never got hybrid (or pure nvidia) mode working properly with suspend & resume on this laptop though.

**Comment by skuz-o** [comment]
huge thanks @Askannz for you work. i've been using optimus-manager for years on my one and only Lenovo laptop carrying the ancient GTX950, pushing it to the limit for oh so many hours of 3D gaming fun, while still allowing me to have a good battery time for work unplugged. never had much luck with hybrid mode but being able to switch between GPUs with a simple relog has been a lifesaver and enabled me to use Arch all this time. i have almost zero programming knowledge so won't be picking up the slack, but at least know your time and effort has been greatly appreciated!

**Comment by michaldybczak** [comment]
Thank you for optimus-manager. I've been using it for years, and it worked very well. With time, the old laptop became more stationary, so battery savings are not so important anymore, and hybrid session worked quite well, even on Wayland. So I decided to retire optimus-manager and stay on hybrid session permanently. However, it turned out, that without optimus-manager, multimonitor setup wasn't working, both on X11 and Wayland. With optimus-managager, all worked. I tried to emulate the optimus-manager setup, using its hybrid configs, the same packages, etc. Nothing. Apparently, there is a kernel parameter needed for it to work, but the graphical session won't start with it (black screen), which is a known bug, that was fixed very recently, and I'm still waiting for it to arrive. Somehow, with optimus-manager it was working like a charm, without this kernel parameter, so I keep using as long as it is needed. It literally saved me.
It's sad to hear that the project is retired, but I understand, things must come to an end sometimes and old hardware is used less and less. After a couple of years, optimus-manager probably won't be needed anymore, but at this very moment, it's still very helpful, even if it's for multimonitor-support only.
Thank you!

**Comment by es20490446e** [comment]
When opening a pull request just make sure to mention me: @es20490446e

**Comment by MrDiver** [comment]
Thank you <3

**Comment by ZYZ64738** [comment]
> All pending pull requests have been merged and tested 👍
> 
> From now on each merge is a stable release. So if you are using the AUR, rather install the git package.

Hi - thx for your work. But for me int's not clear how to install the package without using AUR. In the README its mentioned

"Install the optimus-manager package. In the AUR ..."

If I use the AUR-Version my system gets stuck while booting - so I want to try this updated version but I have no clue howto install it.

**Comment by es20490446e** [comment]
In the application "Terminal" enter:
```
curl -s https://raw.githubusercontent.com/Askannz/optimus-manager/master/package/install.sh | bash
```

**Comment by nwildner** [comment]
Just to add some more information to this thread as well. @Askannz added me as a contributor to this repository and I'll be doing the "Support Engineer" role here, pinging folks from long standing issues that are lingering or don't reproduce anymore, to check if the initial reporter still has interest on the case or if the case can't be reproduced anymore because the issue was solved.

This is to keep the queue in a sane state without a lot of backlog.

Feel free to reopen cases that could be relevant to you but please, do not necrobump only to ask random stuff.

**Comment by es20490446e** [comment]
Glad to have you onboard 👍

**Comment by es20490446e** [comment]
`optimus-manager-git` on the AUR works fine again.

**Comment by es20490446e** [comment]
Since now every commit is tested for stability, the regular `optimus-manager` package in the AUR is no longer needed. Hence I have requested its removal.

If you are using that one, switch to `optimus-manager-git`.

**Comment by nwildner** [comment]
Quick update. Last 7 days he had [55 cases closed](https://github.com/Askannz/optimus-manager/pulse). Reasons involved:

- Original reported has no way to reproduce it again(does not have the laptop or isn't using `optimus-manager` anymore)
- Original reported timed out or it is not interested on the case for any other reason
- Cases being merged into one main case(nvidia black screen issue, wayland support issue...) 
- Features that are way to specific (like the thinkfan one) where other projects have more power to fix that and `optimus-manager` will have to implement a circumvent mechanism to deal with the case, or some very specific technology like the `Ly` login manager.
  - We're more than happy to accept PRs here. We just don't have the workforce(yet) to develop such specific features here.
- Driver way too old (`nvidia-450`)  and `optimus-manager` version from 3+ years ago or any other component that is way too old for us to reproduce again and that the issue might be gone already.

So far, reporters and contributors have been very receptive with the fact we're making some clean up here to put the project back on track.

I'm still monitoring some issues where the reported didn't came back, and I'll close them on the next week or so.

**Comment by es20490446e** [comment]
As said in https://github.com/Askannz/optimus-manager/issues/452, we have now very simple Wayland support.

If you experience any issue, let us know 👍

**Comment by es20490446e** [comment]
Probably we could improve desktop animations by setting the min frequency of the GPU a bit higher.

Discussion [here](https://github.com/NVIDIA/open-gpu-kernel-modules/discussions/714).

**Comment by es20490446e** [comment]
All actionable known issues have been fixed 🔧

**Comment by nwildner** [comment]
Hello folks.

Just dropping a message that I'm currently not really participating anymore here after the purge we've made on old cases because I've migrated from `i3wm` to `hyprland` and using the default Prime mechanism without `optimus-manager` is working "good enough" and I don't feel I need to use Optimus Manger again.

Nevertheless, thanks for this amazing piece of software that allowed me to correctly run X11 back in the 6.0 kernel days where my Asus laptop was really problematic and `optimus-manager` was the only software that allowed me to have a working i3wm session with some nice battery savings.

Wayland became better on Nvidia and I feel I don't need to completely shutdown my dGPU entirely to have the same amount of battery time I had on X11.

**Comment by es20490446e** [comment]
Since I'm on Plasma, and Plasma is about to drop support for X11, shortly I won't no longer be able to test how Optimus Manager works on X11, and likely not able to fix bugs.

If someone wants to do so, let @Askannz know, so he can add that person as co-maintainer.

Or just click on the "watch" button, on the top right corner of the repository's main screen. To follow new issues.

Anyways I will be addressing timely any pull request that anyone makes.

**Comment by es20490446e** [comment]
If you are only using a Wayland session, you can use [`prime-all`](https://gitlab.com/es20490446e/prime-all).
